### PR TITLE
Support loading and querying const tables

### PR DESCRIPTION
**Pull request by xuebinsu**
As the first part of the pandas-like API, this patch adds support
for loading Python Iterables to Greenplum as "constant tables"
and querying them. This is implemented by using the VALUES
clause in SQL. This is the simplest Table that I can think of.

This patch also adds some minimal test cases using pytest.

### CODE REVIEW

**Review by yihong0618 — APPROVED** [review]
LGTM

File: greenplumpython/table.py [hunk 1]
```diff
@@ -35,28 +30,93 @@ def __getitem__(self, key):
...
-    def name(self):
-        raise NotImplementedError()
 
-    # If all = false, use cursor to fetch
-    def fetch(all: bool = true):
-        raise NotImplementedError()
+def values(rows: Iterable[Tuple], db: db.Database) -> Table:
+    return Table(
+        f"VALUES {','.join(['(' + ','.join(str(datum) for datum in row) + ')' for row in rows])}",
+        db=db,
```
Comment (beeender) [inline greenplumpython/table.py:121]:
```
gpadmin=# with a as ( VALUES (1, 2),(2),(3) ) select * from a;
ERROR:  VALUES lists must all be the same length
LINE 1: with a as ( VALUES (1, 2),(2),(3) ) select * from a;
```

And type checking is necessary.
I can create some tracker stories for those.

**Review by beeender — COMMENTED** [review]
:tada: We are moving forward!

File: greenplumpython/db.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+from typing import Iterable
+
+import psycopg2
+import psycopg2.extras
+
+
+class Database:
+    def __init__(self, **params) -> None:
+        self.conn = psycopg2.connect(
```
[the remark this answers is not in the document]
Comment (xuebinsu) [reply greenplumpython/db.py:9]:
```suggestion
        self._conn = psycopg2.connect(
```

File: greenplumpython/table.py [hunk 3]
```diff
@@ -1,29 +1,24 @@
...
-        self.query = query
-        self.parents = parents
-        if table_name is not None:
-            self.name = table_name
-        else:
-            # FIXME: short UUID?
-            self.name = "tb_" + "".join(random.choice(string.ascii_lowercase) for _ in range(60))
+        self._query = query
+        self._parents = parents
+        self._name = "cte_" + uuid4().hex if name is None else name
```
[the remark this answers is not in the document]
Comment (xuebinsu) [reply greenplumpython/table.py:17]: The `_name` is really the name of the relation if the `Table` is a relation recorded in `pg_class`. In that case, the `_name` is used for retrieving its info (e.g., columns).

For CTEs, the `_name` is only a unique "id" and the user need NOT to be aware of it.

File: greenplumpython/table.py [hunk 4]
```diff
@@ -1,29 +1,24 @@
...
         if any(parents):
-            self.db = next(iter(parents)).db
+            self._db = next(iter(parents))._db
         else:
-            self.db = db
+            self._db = db
```
[the remark this answers is not in the document]
Comment (xuebinsu) [reply greenplumpython/table.py:21]: if `self._db` is None, then the whole query cannot be executed. If the user tries to do so, an error will be thrown automatically. 

We can defintely make the check happen earlier.

File: greenplumpython/table.py [hunk 5]
```diff
@@ -35,28 +30,93 @@ def __getitem__(self, key):
...
             return self.select(key)
-        if isinstance(key, Expr):
+        if isinstance(key, expr.Expr):
             return self.filter(key)
         if isinstance(key, slice):
             raise NotImplementedError()
 
-    def filter(expr):
-        return Table(f"SELECT * FROM {self.name} WHERE {str(key)}", parents=[self])
+    def filter(self, expr: expr.Expr) -> "Table":
```
[the remark this answers is not in the document]
Comment (xuebinsu) [reply greenplumpython/table.py:41]:
```suggestion
    # FIXME: Add test
    def filter(self, expr: expr.Expr) -> "Table":
```

File: greenplumpython/table.py [hunk 6]
```diff
@@ -35,28 +30,93 @@ def __getitem__(self, key):
...
+        return Table(f"SELECT * FROM {self._name} WHERE {str(expr)}", parents=[self])
 
-    def select(self, expr_list: list):
+    def select(self, target_list: Iterable) -> "Table":
```
Comment (xuebinsu) [inline greenplumpython/table.py:44]:
```suggestion
    # FIXME: Add test
    def select(self, target_list: Iterable) -> "Table":
```

File: greenplumpython/table.py (hunk 1 above)
Comment (xuebinsu) [reply greenplumpython/table.py:121]: Agree.

File: greenplumpython/table.py [hunk 7]
```diff
@@ -35,28 +30,93 @@ def __getitem__(self, key):
...
+
+    def name(self) -> str:
+        return self._name
+
+    def _list_lineage(self) -> Iterable["Table"]:
+        lineage = [self]
+        tables_visited = set()
+        current = 0
+        while current < len(lineage):
+            for table in lineage[current].parents:
```
[the remark this answers is not in the document]
Comment (xuebinsu) [reply greenplumpython/table.py:73]:
```suggestion
            for table in lineage[current]._parents:
```

File: greenplumpython/table.py [hunk 8]
```diff
@@ -35,28 +30,93 @@ def __getitem__(self, key):
...
+            f"""
+                SELECT {','.join([str(target) for target in target_list])} 
+                FROM {self._name}
+            """,
+            parents=[self],
         )
 
-    def describe(self):
-        raise NotImplementedError()
+    def columns(self) -> "Table":
```
Comment (xuebinsu) [inline greenplumpython/table.py:53]:
```suggestion
    def column_names(self) -> "Table":
```

File: greenplumpython/table.py [hunk 9]
```diff
@@ -35,28 +30,93 @@ def __getitem__(self, key):
...
+        if any(self._parents):
+            raise NotImplementedError()
+        return Table(
+            f"""
+                SELECT *
```
Comment (xuebinsu) [inline greenplumpython/table.py:58]:
```suggestion
                SELECT column_name
```

File: greenplumpython/table.py [hunk 10]
```diff
@@ -35,28 +30,93 @@ def __getitem__(self, key):
...
+        return Table(
+            f"""
+                SELECT * 
+                FROM information_schema.columns 
+                WHERE table_name = quote_ident('{self._name}')
+            """,
+            db=self._db,
+        )
+
+    def name(self) -> str:
```
[the remark this answers is not in the document]
Comment (xuebinsu) [reply greenplumpython/table.py:65]:
```suggestion
    @property
    def name(self) -> str:
```

File: greenplumpython/table.py [hunk 11]
```diff
@@ -1,29 +1,24 @@
...
 
 
 class Table:
     def __init__(
         self,
         query: str,
         parents: Iterable["Table"] = [],
-        table_name: str = None,
-        db: Database = None,
+        name: str = None,
```
[the remark this answers is not in the document]
Comment (beeender) [reply greenplumpython/table.py:12]: Yup, I will do that.

File: greenplumpython/table.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (beeender) [reply greenplumpython/table.py:17]: Consider this case:

```
t1 = Table("SomeTable")
t2 = Table("SomeTable") # same table

t3 = t1.join(t2, t2['id'] = t1['id']
```

Will it be a problem for the `_list_lineage()` implementation?

P.S. We don't have to solve the problem for now. Just a reminder for future implementation.

**Review by beeender — APPROVED** [review]

**Review by Quasimewdo — COMMENTED** [review]
Finished Review

**Pull request merged**